File: RailSDK-Python/rail/discovery.py
# ============================================================================
# METHOD DISCOVERY
# ============================================================================
# Auto-discovery of methods using Python introspection.
# Generates manifest JSON for LLM consumption.
#
# ============================================================================

import inspect
import typing
from typing import List, Dict, Any, Optional, get_type_hints

from .types import FunctionInfo, ParameterInfo

# ============================================================================
# TYPE MAPPING
# ============================================================================

_TYPE_MAP = {
    int: "integer",
    float: "number",
    str: "string",
    bool: "boolean",
    list: "array",
    dict: "object",
    type(None): "null",
}

def _python_type_to_json(py_type: Any) -> str:
    """Convert Python type to JSON Schema type."""
    # Handle None/Optional first
    if py_type is type(None):
        return "null"
    
    # Handle Optional[X] -> X
    origin = getattr(py_type, "__origin__", None)
    if origin is typing.Union:
        args = py_type.__args__
        # Filter out NoneType for Optional handling
        non_none_args = [a for a in args if a is not type(None)]
        if len(non_none_args) == 1:
            return _python_type_to_json(non_none_args[0])
        return "any"
    
    # Handle List[X]
    if origin is list:
        return "array"
    
    # Handle Dict[K, V]
    if origin is dict:
        return "object"
    
    # Handle basic types
    return _TYPE_MAP.get(py_type, "any")

# ============================================================================
# DISCOVERY FUNCTIONS
# ============================================================================

def discover_methods(instance: object, include_private: bool = False) -> List[FunctionInfo]:
    """
    Discover all callable methods on an object instance.
    
    Args:
        instance: Object instance to inspect
        include_private: If True, include methods starting with _
    
    Returns:
        List of function descriptors suitable for manifest
    """
    functions: List[FunctionInfo] = []
    
    # Get the class and its methods
    cls = instance.__class__
    
    for name in dir(instance):
        # Skip magic methods
        if name.startswith("__"):
            continue
        
        # Skip private unless requested
        if name.startswith("_") and not include_private:
            continue
        
        attr = getattr(instance, name)
        
        # Only include callable methods
        if not callable(attr):
            continue
        
        # Skip bound methods from base object class
        if hasattr(object, name):
            continue
        
        # Extract function info
        func_info = _extract_function_info(attr, name)
        if func_info:
            functions.append(func_info)
    
    return functions

def _extract_function_info(func: callable, name: str) -> Optional[FunctionInfo]:
    """Extract function information from a callable."""
    try:
        sig = inspect.signature(func)
        doc = inspect.getdoc(func) or ""
        
        # Try to get type hints
        try:
            hints = get_type_hints(func)
        except Exception:
            hints = {}
        
        parameters: List[ParameterInfo] = []
        
        for param_name, param in sig.parameters.items():
            # Skip 'self' for methods
            if param_name == "self":
                continue
            
            param_type = hints.get(param_name, inspect.Parameter.empty)
            
            param_info: ParameterInfo = {
                "name": param_name,
                "type": _python_type_to_json(param_type) if param_type != inspect.Parameter.empty else "any",
                "description": "",  # Could extract from docstring parsing
                "required": param.default == inspect.Parameter.empty
            }
            
            parameters.append(param_info)
        
        return {
            "name": name,
            "description": doc.split("\n")[0] if doc else "",  # First line of docstring
            "parameters": parameters
        }
    
    except Exception:
        # If we can't inspect, skip this method
        return None

# ============================================================================
# MANIFEST GENERATION
# ============================================================================

def generate_manifest(
    instance: object,
    context: Optional[str] = None,
    include_private: bool = False
) -> Dict[str, Any]:
    """
    Generate a complete manifest for an object instance.
    
    Args:
        instance: Object to generate manifest for
        context: Optional context name
        include_private: If True, include private methods
    
    Returns:
        Complete manifest dictionary
    """
    import os
    from .version import __version__
    
    functions = discover_methods(instance, include_private)
    
    return {
        "processId": os.getpid(),
        "language": "python",
        "sdkVersion": __version__,
        "context": context or instance.__class__.__name__,
        "functions": functions
    }

File: RailSDK-Node/src/discovery.ts
/**
 * ============================================================================
 * METHOD DISCOVERY
 * ============================================================================
 * Auto-discovery of methods using JavaScript/TypeScript introspection.
 * Generates manifest JSON for LLM consumption.
 *
 * ============================================================================
 */

import { FunctionInfo, ParameterInfo, Manifest } from './types';
import { VERSION } from './version';

// ============================================================================
// TYPE MAPPING
// ============================================================================

/**
 * Map common JavaScript types to JSON Schema types.
 */
function inferType(value: unknown): string {
    if (value === null) return 'null';
    if (typeof value === 'undefined') return 'any';
    if (typeof value === 'string') return 'string';
    if (typeof value === 'number') return 'number';
    if (typeof value === 'boolean') return 'boolean';
    if (Array.isArray(value)) return 'array';
    if (typeof value === 'object') return 'object';
    return 'any';
}

// ============================================================================
// PARAMETER EXTRACTION
// ============================================================================

/**
 * Extract parameter names from a function.
 * Uses function.toString() parsing.
 */
function extractParams(fn: Function): ParameterInfo[] {
    const fnStr = fn.toString();

    // Match function parameters: function(a, b, c) or (a, b, c) =>
    const match = fnStr.match(/(?:function\s*\w*\s*)?\(([^)]*)\)/);
    if (!match || !match[1]) return [];

    const paramStr = match[1].trim();
    if (!paramStr) return [];

    const params: ParameterInfo[] = [];

    for (const part of paramStr.split(',')) {
        const trimmed = part.trim();
        if (!trimmed) continue;

        // Handle destructuring, defaults, etc.
        let name = trimmed
            .split(':')[0]  // Remove type annotation (TypeScript)
            .split('=')[0]  // Remove default value
            .trim();

        // Skip rest parameters marker
        if (name.startsWith('...')) {
            name = name.slice(3);
        }

        params.push({
            name,
            type: 'any',
            description: '',
            required: !trimmed.includes('=')  // Has default = optional
        });
    }

    return params;
}

// ============================================================================
// DISCOVERY FUNCTIONS
// ============================================================================

/**
 * Discover all callable methods on an object instance.
 */
export function discoverMethods(instance: object, includePrivate: boolean = false): FunctionInfo[] {
    const functions: FunctionInfo[] = [];

    // Get prototype methods
    const proto = Object.getPrototypeOf(instance);
    const methodNames = Object.getOwnPropertyNames(proto);

    for (const name of methodNames) {
        // Skip constructor and special methods
        if (name === 'constructor') continue;
        if (name.startsWith('__')) continue;
        if (name.startsWith('_') && !includePrivate) continue;

        const method = (instance as any)[name];
        if (typeof method !== 'function') continue;

        const funcInfo: FunctionInfo = {
            name,
            description: '',  // Could extract from JSDoc if available
            parameters: extractParams(method)
        };

        functions.push(funcInfo);
    }

    // Also get own properties that are functions
    for (const name of Object.keys(instance)) {
        if (name.startsWith('_') && !includePrivate) continue;

        const method = (instance as any)[name];
        if (typeof method !== 'function') continue;

        // Skip if already added from prototype
        if (functions.some(f => f.name === name)) continue;

        const funcInfo: FunctionInfo = {
            name,
            description: '',
            parameters: extractParams(method)
        };

        functions.push(funcInfo);
    }

    return functions;
}

/**
 * Generate a complete manifest for an object instance.
 */
export function generateManifest(
    instance: object,
    context?: string,
    includePrivate: boolean = false
): Manifest {
    const functions = discoverMethods(instance, includePrivate);

    return {
        processId: process.pid,
        language: 'node',
        sdkVersion: VERSION,
        context: context || instance.constructor.name,
        functions
    };
}


